django/db/models/sql/datastructures.py

```diff
@@ -114,17 +114,28 @@ def relabeled_clone(self, change_map):
             self.join_field, self.nullable, filtered_relation=filtered_relation,
         )
 
-    def equals(self, other, with_filtered_relation):
+    @property
+    def identity(self):
         return (
-            isinstance(other, self.__class__) and
-            self.table_name == other.table_name and
-            self.parent_alias == other.parent_alias and
-            self.join_field == other.join_field and
-            (not with_filtered_relation or self.filtered_relation == other.filtered_relation)
+            self.__class__,
+            self.table_name,
+            self.parent_alias,
+            self.join_field,
+            self.filtered_relation,
         )
 
     def __eq__(self, other):
-        return self.equals(other, with_filtered_relation=True)
+        if not isinstance(other, Join):
+            return NotImplemented
+        return self.identity == other.identity
+
+    def __hash__(self):
+        return hash(self.identity)
+
+    def equals(self, other, with_filtered_relation):
+        if with_filtered_relation:
+            return self == other
+        return self.identity[:-1] == other.identity[:-1]
```

Comment:
I did a little grepping:
```
django/db/models/sql/datastructures.py:117:    def equals(self, other, with_filtered_relation):
django/db/models/sql/datastructures.py:123:            (not with_filtered_relation or self.filtered_relation == other>
django/db/models/sql/datastructures.py:127:        return self.equals(other, with_filtered_relation=True)
django/db/models/sql/datastructures.py:163:    def equals(self, other, with_filtered_relation):
django/db/models/sql/query.py:962:                if a in reuse and j.equals(join, with_filtered_relation=False)
```
There was only one use of `with_filtered_relation=True` in `Join.__eq__()` which is being removed in this PR. In that case, we could just drop the `with_filtered_relation` argument, always assuming the `False` case:
```suggestion
    def equals(self, other):
        # Ignore filtered_relation in the identity check.
        return self.identity[:-1] == other.identity[:-1]
```

(If we don't remove this, maybe we could make `with_filtered_relation` keyword-only.)

Reply:
I'm open to making this change but this PR has little to do with this particular function, I just happened to refactor it to use `identity` since I needed to implement `__eq__` and `__hash__` to allow `Query.__eq__` and `__hash__` to work.